# 01 - Data Types and Descriptive Statistics


## 1) Learning Objectives

- Identify qualitative and quantitative variables.
- Inspect a dataset correctly before analysis.
- Compute and interpret descriptive statistics.
- Compare groups with simple summaries.
- Connect summary tables to real analysis decisions.


## 2) Why This Topic Matters

If we misunderstand variable types, every later step can become wrong.

In real projects, many mistakes come from:

- treating category labels as numeric measurements,
- skipping data inspection,
- reading numbers without interpretation.

This notebook builds a safe and repeatable starting workflow.


## 3) Short and Simple Theory Explanation

- **Qualitative variable**: category label (for example, species).
- **Quantitative variable**: numeric measurement (for example, sepal length).
- **Descriptive statistics** summarize a dataset using:
  - center (mean, median),
  - spread (standard deviation, min, max),
  - simple shape clues.


## 4) Step-by-Step Code Examples

### Step 4.1 - Import libraries and set style

**What are we trying to learn?**  
How to prepare a clean and reproducible notebook environment.

**What should we expect?**  
No major output. This is setup for all later steps.


In [1]:
# Numeric work
import numpy as np
# Data tables
import pandas as pd
# Plotting
import seaborn as sns
import matplotlib.pyplot as plt

# Iris dataset
from sklearn.datasets import load_iris

# Reproducibility for any random operations
np.random.seed(42)

# Consistent visual style
sns.set_theme(style="whitegrid")


**Interpretation**

This setup keeps results consistent and readable.

**Why this matters**

Reproducibility is essential in academic work.

**Common mistake**

Skipping random seed and getting different results each run.


### Step 4.2 - Load Iris and create readable class labels

**What are we doing?**  
Load the dataset and replace numeric target codes with species names.

**What should students notice?**  
The final table has four numeric columns and one label column.


In [2]:
# Load Iris with pandas-ready frame
iris = load_iris(as_frame=True)

# Copy so we can edit safely
df = iris.frame.copy()

# Map numeric target values (0,1,2) to names
label_map = dict(enumerate(iris.target_names))
df["species"] = df["target"].map(label_map)

# Drop numeric target after creating labels
df = df.drop(columns=["target"])

# Quick first look
print(f"Shape: {df.shape}")
df.head()


Shape: (150, 5)


,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),species
0,5.1,3.5,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,3.1,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa


**Interpretation**

- We have 150 rows and 5 columns.
- Four features are numeric measurements.
- `species` is a category label.

**Why this matters**

Readable labels reduce interpretation mistakes in tables and plots.


### Step 4.3 - Inspect structure with `info()` and `describe()`

**What are we trying to learn?**  
How to validate data types, missing values, and quick summaries.

**What should we expect?**

- Numeric columns as float type,
- no missing values,
- basic center/spread statistics.


In [3]:
# Check dtypes and missing values
df.info()

# Summarize numeric and categorical columns
df.describe(include="all").T


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   species            150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
sepal length (cm),150.0,NaN,NaN,NaN,5.843333,0.828066,4.3,5.1,5.8,6.4,7.9
sepal width (cm),150.0,NaN,NaN,NaN,3.057333,0.435866,2.0,2.8,3.0,3.3,4.4
petal length (cm),150.0,NaN,NaN,NaN,3.758,1.765298,1.0,1.6,4.35,5.1,6.9
petal width (cm),150.0,NaN,NaN,NaN,1.199333,0.762238,0.1,0.3,1.3,1.8,2.5
species,150,3,setosa,50,NaN,NaN,NaN,NaN,NaN,NaN,NaN


**Interpretation**

This confirms the dataset is clean and ready for descriptive analysis.

**Common mistake**

Jumping to plots or tests before checking missing values and dtypes.


### Step 4.4 - Identify variable types explicitly

**What are we doing?**  
Create clear lists for qualitative and quantitative variables.


In [4]:
# Category-like columns
qualitative_vars = ["species"]

# Numeric columns
quantitative_vars = [col for col in df.columns if col not in qualitative_vars]

print("Qualitative:", qualitative_vars)
print("Quantitative:", quantitative_vars)


Qualitative: ['species']
Quantitative: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']


**Interpretation**

Variable type controls which statistics and visuals are valid.

**Why this matters**

A mean for a category label is not meaningful.


### Step 4.5 - Grouped descriptive statistics by species

**What should we expect?**  
Species groups should differ in some flower measurements.


In [5]:
# Group-wise summary table for central tendency and spread
group_summary = (
    df.groupby("species")[quantitative_vars]
      .agg(["mean", "median", "std", "min", "max"])
)

group_summary


sepal length (cm)                            sepal width (cm)  \
                        mean median       std  min  max             mean   
species                                                                    
setosa                 5.006    5.0  0.352490  4.3  5.8            3.428   
versicolor             5.936    5.9  0.516171  4.9  7.0            2.770   
virginica              6.588    6.5  0.635880  4.9  7.9            2.974   

                                      petal length (cm)                        \
           median       std  min  max              mean median       std  min   
species                                                                         
setosa        3.4  0.379064  2.3  4.4             1.462   1.50  0.173664  1.0   
versicolor    2.8  0.313798  2.0  3.4             4.260   4.35  0.469911  3.0   
virginica     3.0  0.322497  2.2  3.8             5.552   5.55  0.551895  4.5   

                petal width (cm)                             
            max             mean median       std  min  max  
species                                                      
setosa      1.9            0.246    0.2  0.105386  0.1  0.6  
versicolor  5.1            1.326    1.3  0.197753  1.0  1.8  
virginica   6.9            2.026    2.0  0.274650  1.4  2.5

**Interpretation**

Group summaries reveal class-level patterns that overall averages can hide.

**Why this matters**

This is often the first clue for useful features in later ML tasks.


### Step 4.6 - Small manual intuition example (from old materials style)

**What are we doing?**  
Use a tiny list and compute simple summaries directly.

**Why**  
Manual steps strengthen intuition before package automation.


In [6]:
# Small list used for intuition practice
small_data = [5, 7, 8, 7, 10, 6, 9, 8, 7, 6]

# Manual mean
manual_mean = sum(small_data) / len(small_data)

# Manual median
sorted_data = sorted(small_data)
n = len(sorted_data)
manual_median = (sorted_data[n//2 - 1] + sorted_data[n//2]) / 2

print("Data:", small_data)
print("Manual mean:", manual_mean)
print("Manual median:", manual_median)


Data: [5, 7, 8, 7, 10, 6, 9, 8, 7, 6]
Manual mean: 7.3
Manual median: 7.0


**Interpretation**

Manual calculations show what packages are computing behind the scenes.

**Common mistake**

Using package output without understanding its meaning.


## 8) Small Concept Check Questions (with answers)

**Q1:** Why must we separate qualitative and quantitative variables?  
**A1:** Because they need different summaries and visualizations.

**Q2:** Why use grouped summaries by species?  
**A2:** They reveal differences that overall means may hide.


## 9) Key Takeaways

- Inspect before interpreting.
- Variable types are foundational.
- Grouped descriptive statistics are more informative than single global numbers.
- Manual examples improve statistical intuition.


## 10) Optional Short Practice / Challenge

- Convert `species` to categorical dtype.
- Recompute grouped stats and compare memory usage before/after conversion.
